Use multiple solvers in qalpha

edenfrenkel opened the pull request:
This defines a `BasicSolver` trait in `solver/src/basics.rs`, which allows making very basic `check-sat` queries which return a `sat` result together with a trace, an `unsat` result together with an unsat-core, or `unknown`. By abstracting this basic functionality it is possible to implement this trait for a single `SolverConf`, as well as for a vector of `SolverConf`'s in two different ways, once as `FallbackSolvers` tried sequentially and second as `ParallelSolvers` tried in parallel. This is utilized in `qalpha`, which now uses `ParallelSolvers` for its inductiveness and safety checks by default, and is configurable to use `FallbackSolvers` instead via `--fallback`. The simulation queries in `qalpha` still use a single `SolverConf`, now abstracted behind the `BasicSolver` trait as well.

Solver cancellation has also been moved out of the inference code and made more generic, using the new `BasicSolverCanceler` trait and `SolverCancelers` struct, which aggregates multiple `BasicSolverCanceler`'s and itself implements `BasicSolverCanceler`. This is used recursively to create hierarchical tree-like cancellation, where the cancellation of a node cancels all of its descendants. The idea is that functions should only cancel queries which they (perhaps recursively) launched and not assume anything about the behavior of their callers. When the result is bubbled up to those callers, they should cancel their queries according to their own desired behavior. This allows more general applicablity which is independent of the specific goals of any single use-case, (in this case `qalpha`, which cancels all solvers on the first `sat` result.)

Another change is that now transition queries parallelize over disjunctive transitions, with proper cancellation if a `sat` response is encountered.

tchajed commented on inference/src/basics.rs:
```diff
@@ -46,24 +72,15 @@ impl<'a> Core<'a> {
-                !self.participants.contains(i) && counter_model.eval(&self.formulas[*i]) == 0
-            });
+        let new_participant = (0..self.formulas.len()).find(|i| {
+            !self.participants.contains(i) && counter_model.eval(&self.formulas[*i]) == 0
```
You can write this using `self.formulas.iter().enumerate()`.

tchajed commented on inference/src/basics.rs:
```diff
@@ -377,27 +453,35 @@ impl FOModule {
+            match try_conf(confs[conf_idx], &mut core) {
+                CexResult::Cex(model) => {
+                    if !core.add_counter_model(model[0].clone()) {
+                        return CexResult::Cex(model);
```
Can you add a comment here about the control flow? There also isn't a comment on `add_counter_model` explaining what the boolean return value is.

tchajed commented on solver/src/backends.rs:
```diff
@@ -69,6 +69,16 @@ impl GenericBackend {
     }
+
+    /// Get the solver type.
+    pub fn get_solver_type(&self) -> SolverType {
```
You can name this `fn solver_type`.

tchajed commented on solver/src/backends.rs:
```diff
@@ -69,6 +69,16 @@ impl GenericBackend {
+    }
+
+    /// Get the solver timeout.
+    pub fn get_timeout_ms(&self) -> Option<usize> {
```
You can name this `fn timeout_ms`

edenfrenkel replied:
This is the only comment I couldn't address because there's already such a function which sets the timeout value.

tchajed commented on solver/src/conf.rs:
```diff
@@ -24,4 +38,50 @@ impl SolverConf {
+        backend_type: SolverType,
+        smt: bool,
+        fname: &String,
+        timeout: usize,
```
Please name this argument `timeout_s` to avoid units confusion.

tchajed commented on solver/src/conf.rs:
```diff
@@ -3,11 +3,25 @@
-use crate::{backends::GenericBackend, imp::Solver};
 use fly::syntax::Signature;
 
+fn solver_default_bin(t: SolverType) -> &'static str {
```
You can make this a method of `SolverType` and name it `fn bin_name`.

tchajed commented on inference/src/basics.rs:
```diff
@@ -209,164 +258,191 @@ impl FOModule {
         };
 
-        loop {
+        let try_conf = |conf: &SolverConf, core: &mut Core| -> CexResult {
```
This is actually a core piece of (potentially) general functionality for working with a "meta solver conf" that consists of a sequence of configurations to try. If you can it would be great to factor this out as much as possible. In this PR I'd definitely want this function to be separated into its own function with documentation. Eventually it would be great if we could somehow wrap `Vec<SolverConf>` in a struct with an API similar to a single `SolverConf` (perhaps `Solver` needs to be a trait for this to work).

tchajed commented on inference/src/fixpoint.rs:
```diff
@@ -115,23 +115,32 @@ impl FoundFixpoint {
+        SolverConf::new(SolverType::Z3, true, &infer_cfg.fname, 0, 7),
+    ];
+    let confs = solver_confs.iter().collect_vec();
+    let simulation_conf = SolverConf::new(SolverType::Z3, true, &infer_cfg.fname, 3, 0);
```
Please move all of this policy configuration to a separate function (even if it just returns the constants for `solver_confs` and `simulation_conf`) and add some documentation on why you made these choices. It could also use a reminder as to what those two integer arguments to `SolverConf::new` are.

tchajed reviewed (commented):
I left a bunch of comments, but it's mostly just code-style advice.

This PR would be a great candidate for benchmarking, so you might want to add some benchmarking on top of #142. If you already have some data on the performance impact of this change from your own benchmarking script please include that data in the PR description.

edenfrenkel commented:
@tchajed When trying to factor out `try_conf` I realized I should probably just implement multi-solvers more generally using a solver trait, like you said. This turned out to be more tricky than expected, unless I reduced solver behavior to something very basic, which is now implemented in the `BasicSolver` trait in `solver/src/basics.rs`. This trait allows making very simple `check_sat` queries which return sat+trace, unsat+unsat-core, or unknown. I implemented this trait for a single `SolverConf`, as well as for a vector of `SolverConf`'s **twice**, first as fallback solvers and second as parallel solvers. Solver cancellation has also been moved out of the inference code and made more generic.

Let me know what you think. I'll rebase and update the pull request description soon.

tchajed reviewed (approved):
I like the trait solution, that nicely abstracts over what we need in this case!

Can you update the PR description? Other than that (and the usual rebase onto main) this is ready to merge.

edenfrenkel commented:
@tchajed updated and rebased
